### ISSUE TITLE
Non-internal version of HexStringToBytesInternal

### ISSUE THREAD

**Issue by g-easy** [abseil/abseil-cpp#234]
Please consider making HexStringToBytesInternal and BytesToHexStringInternal not internal. It would be good to have a non-allocating alternative to absl::HexStringToBytes, for performance.

**Comment by mbxx** [comment]
I suspect this is something that we're going to want to keep internal, but it might help to know a little more about your situation.  Did you run into a specific situation in which the extra allocations from `absl::HexStringToBytes` are causing a problem?  Or are you requesting a non-allocating alternative for API completeness?

When we first released Abseil, we considered open sourcing a non-allocating version of `absl::HexStringToBytes` but decided against it.  There's a bit of a tension that we find ourselves running into constantly: we've promised not to break users of our code but we also need the ability to keep making changes to our library.  One of the ways we've handled this is to keep our public API surface as small as possible.  To that end, we've open sourced the most widely used pieces of low level code used at Google, but left out some of the rarely used pieces because they would increase the size of the API that we need to support without providing a lot of additional value to our users.  The non-allocating `absl::HexStringToBytes` fell into the "rarely used" bucket.

https://github.com/abseil/abseil-cpp/issues/141 also throws a bit of a monkey wrench into releasing a non-allocating `absl::HexStringToBytes`.  I'd like to address that issue by introducing something like:

```
ABSL_MUST_USE_RESULT bool HexStringToBytes(absl::string_view from, string* to);
```

which is basically the right signature for a non-allocating overload, but this overload would also do input validation, which might not be what you want if you're focused on performance.  If we wanted to introduce a new function for performance, we'd likely want to give it a new name.

So I suspect we're not going to want to add a public version of `HexStringToBytesInternal`, but we would find any specific performance issues that you're running into very interesting.
